src/services/sourcemaps.ts

```diff
@@ -0,0 +1,330 @@
+/* @internal */
+namespace ts.sourcemaps {
+    export interface SourceMapData {
+        version?: number;
+        file?: string;
+        sourceRoot?: string;
+        sources: string[];
+        sourcesContent?: string[];
+        names?: string[];
+        mappings: string;
+    }
+
+    export interface SourceMappableLocation {
+        fileName: string;
+        position: number;
+    }
+
+    export interface SourceMapper {
+        getOriginalPosition(input: SourceMappableLocation): SourceMappableLocation;
+        getGeneratedPosition(input: SourceMappableLocation): SourceMappableLocation;
+    }
+
+    export const identitySourceMapper = { getOriginalPosition: identity, getGeneratedPosition: identity };
+
+    export interface SourceMapDecodeHost {
+        readFile(path: string): string;
+        fileExists(path: string): boolean;
+        getCanonicalFileName(path: string): string;
+        log(text: string): void;
+    }
+
+    export function decode(host: SourceMapDecodeHost, mapPath: string, map: SourceMapData, program?: Program, fallbackCache = getSourceFileLikeCache(host)): SourceMapper {
+        const currentDirectory = getDirectoryPath(mapPath);
+        const sourceRoot = map.sourceRoot || currentDirectory;
+        let decodedMappings: ProcessedSourceMapSpan[];
+        let forwardSortedMappings: ProcessedSourceMapSpan[];
+        let reverseSortedMappings: ProcessedSourceMapSpan[];
+
+        return {
+            getOriginalPosition,
+            getGeneratedPosition
+        };
+
+        function getGeneratedPosition(loc: SourceMappableLocation): SourceMappableLocation {
+            const maps = filter(getForwardSortedMappings(), m => comparePaths(loc.fileName, m.sourcePath, sourceRoot) === 0);

```

Comment:
Couldn't we `binarySearch` without the `filter` and then check whether the result points to a mapping in a different file? `filter` requires a full scan of the array prior to using the more efficient `binarySearch`.